### PR TITLE
Fix sort for slow gradcheck

### LINKED ISSUE #57166: gradcheck fails for slow mode but succeeds for fast mode for 'sort' and 'msort' on large sample inputs
To replicate:
1. Update sort's OpInfo to have `gradcheck_fast_mode=False`
3. run `python test/test_ops.py -k sort`

edit: The issue may be that the perturbing the inputs by the default value of eps 1e-6 changes its sorted order. Fast mode is succeeding because the Jacobian matrix is extremely large (8000 x 8000) and only a small percentage of elements are different ( because the Jacobian of a sort is a permutation matrix). 

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @heitorschueroff

### PR DESCRIPTION
**Pull request by soulitzer**
Fixes #57166

Generate new inputs until we get one where we know that x + eps won't change its sorted order.

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -1327,9 +1327,21 @@ def small_3d_unique(dtype, device):
...
+    valid_sample = False
+    # NOTE: Make sure that if you change eps, the eps used by
+    # gradcheck must be updated to be at least this value
+    eps = 1e-6
+    while not valid_sample:
+        largesample = SampleInput(make_tensor((L**3,), device, dtype,
+                                  low=None, high=None,
+                                  requires_grad=requires_grad))
+        # TODO: Maybe we shouldn't assume sort is correct here
+        sorted_input, _ = largesample.input.sort()  # type: ignore[union-attr]
```
Comment (albanD) [inline torch/testing/_internal/common_methods_invocations.py:1341]: You can work directly with the result of `make_tensor` to avoid typing issues.

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -1327,9 +1327,20 @@ def small_3d_unique(dtype, device):
...
-    largesample = SampleInput(make_tensor((L**3,), device, dtype,
-                              low=None, high=None,
-                              requires_grad=requires_grad))
+    # We need to make sure that when we perform gradcheck, perturbing input
+    # by eps does not change the sorted order
+    valid_sample = False
+    # NOTE: Make sure that if you change eps, the eps used by
+    # gradcheck must be updated to be at least this value
+    eps = 1e-6
+    while not valid_sample:
```
Comment (albanD) [inline torch/testing/_internal/common_methods_invocations.py:1336]: This is not usually the best thing as someone changing L would be welcomed with test that run forever :p

Maybe use arange and then divide the result by 100 to get floating point values that are all far enough appart.

Comment (soulitzer) [reply torch/testing/_internal/common_methods_invocations.py]: Ah good catch. I'm going to use randperm so that it comes pre-shuffled

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -2964,10 +2968,20 @@ def generator():
...
 def sample_inputs_msort(op_info, device, dtype, requires_grad):
-    # test large sorting codepath
-    largesample = SampleInput(make_tensor((L**3,), device, dtype,
-                                          low=None, high=None,
-                                          requires_grad=requires_grad))
+    def apply_grad(t):
+        if dtype in floating_types_and(torch.float16, torch.bfloat16):
+            t.requires_grad_(requires_grad)
+
+    def large_1d_unique(dtype, device):
```
Comment (albanD) [inline torch/testing/_internal/common_methods_invocations.py:2975]: Can we move that so that we don't have to have it twice?

**Review by albanD — APPROVED** [review]
Accepting to unblock master.
Would be nice to remove duplication afterwards if possible.
